---
title: "Single Cell RNASeq analysis walkthrough"
author: "Eric J. Kort"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: pygments
    number_sections: true
    css: custom.css
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = TRUE,  message=FALSE, warning=FALSE)
library(monocle3)
library(Seurat)
library(ggplot2)
library(dplyr)

# the following packages are optional
library(cowplot)
library(grid)
```

# Code conventions

Most of the code below is meant to be entered at the R console, and is 
annotated like this:

```{r eval=FALSE}

# Code formatted like this should be pasted into an R session
print("hello world")

```

A few code snippets will need to be entered at the command line (or "Terminal").
Command line code is annotated like this:

```{bash eval=FALSE}

# this is code to be entered at the command line
echo Hello world

```

# Prerequisites {#prerequisites}

This document provides a guided analysis of a scRNASeq dataset. The only 
pre-requisites are access to a "well formed" computer/server with R installed. 
By "well formed" I mean that the computer you are working on will need the 
usual GNU toolchain and related tools (gcc, gfortran, libssl, hdf5 libraries, 
etc). MacOS uses clang and that toolchain is fine too. 

## Ubuntu and other Linux distros

From a fresh installation of Ubuntu 18.04, the following will ensure the 
requisite tools are installed. For other flavors of Linux, the steps should 
be analogous, though not necessarily exactly the same depending on which 
packaging system you are using.

```{bash eval=FALSE}
# switch to root shell
sudo -s

# install build tools
apt update
apt install -y build-essential
apt intall -y gfortran

# The version of R in the ubuntu releases always lags, so we need to add 
# the R-project repository, get its security key, and then install R

# a little shell magic to get our Ubuntu release's "codename"
RELEASE=$(lsb_release -as 2>/dev/null | tail -n1)

echo "deb https://cloud.r-project.org/bin/linux/ubuntu ${RELEASE}-cran40/" >> /etc/apt/sources.list
apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 51716619E084DAB9
apt  update
apt -y install r-base

# the packages we will need have some system level dependencies. 
# so we will install those
apt -y install libssl-dev
apt -y install libcurl4-openssl-dev
apt -y install libxml2-dev

```

Verify your R version is >= 4.0.2 as follow:

```{bash eval=FALSE}
R --version
```

## MacOS

The clang environment provided by Apple should include all the 
dependencies you need except for gfortran which must be installed separately.

To see if you have these tools installed, open a terminal window and enter the 
following commands.

```{bash eval=FALSE}
clang -v
gfortran -v
```

If that returns a version (like `Apple clang version 11.0.0` and 
`gcc version 8.2.0 (GCC)`), you should be all set.

If clang is not found, run the following from the command line:

```{bash, eval=FALSE}
xcode-select --install 
```

If gfortran is not found, [download and install it from here](https://github.com/fxcoudert/gfortran-for-macOS/releases/tag/8.2).
(Open the disk image, then open the ('gfortran-8.2-Mojave' folder and double 
click the "gfortran.pkg" item to run the installer).

If you have not already done so, download and install R and (highly recommended) 
RStudio according to the directions [here](https://rstudio.com/products/rstudio/download/#download).

## Windows

Windows binaries for the necessary tools should be available from CRAN, so 
simply download and install R and (highly recommended) 
RStudio according to the directions [here](https://rstudio.com/products/rstudio/download/#download).

You will not be able to build source-only packages with this setup. But you can
easily install the additional tools needed for building R packages (and even R
itself) on your Windows box. The R team very helpfully provides the necessary
software (based on gcc, mingw, and msys) and instructions for building R and its
packages from source (see instructions
[here](https://cran.r-project.org/bin/windows/Rtools)).

**Note that you do not need to build R from source**, you just need these tools
installed and configured as explained at the link above there to build packages
from source. In particular, there is currently a conflict between the `cpp11`
package and the `tidyr` package that requires building the latest `cpp11`
version on github from source (see below) in order to install `tidyr` (required
by Seurat and, I assume, Monocle as well).

# Installing analysis tools

## Installing Seurat

We will be using the Seurat and Monocle (version 3) packages for the analyses
here. Seurat is available on CRAN so installation should be easy. Try the 
following. If it fails, read the next section. 

```{r eval=FALSE}
install.packages('Seurat')
```
## If Seurat installation fails

While installing Seurat should be easy, at the moment there is a conflict
between the latest release version of the `cpp11` and `tidyr` (on which Seurat
depends). So we need to do a little work around.to install the development
version of `cpp11` that fixes that. **Note:** This requires building from
source, so you will need to have gcc or clang or mingw/msys (Window) installed.
See additional notes in the [prerequisites section above](#prerequisites).

```{r eval=FALSE}
if (!requireNamespace("devtools", quietly = TRUE))
  install.packages("devtools")
devtools::install_github("r-lib/cpp11")
install.packages('Seurat')
```

## Installing Monocle3

The latest Monocle must be installed from github, and has some pre-requisites 
that must be installed. The pre-requisites are part of the Bioconductor 
project so we will add the Bioconductor repositories to the list of searched 
repositories when installing packages.

```{r eval=FALSE}

setRepositories(ind=1:6)
install.packages(c('BiocGenerics', 'DelayedArray', 'DelayedMatrixStats',
                       'limma', 'S4Vectors', 'SingleCellExperiment',
                       'SummarizedExperiment', 'batchelor', 'Matrix.utils'))

# install leidenbase and monocle3 from github
if (!requireNamespace("devtools", quietly = TRUE))
  install.packages("devtools")

# install_github will error out if installed binary packages were built under 
# version of R newer than what we have locally. Change setting to ignore these 
# warnings.
Sys.setenv(R_REMOTES_NO_ERRORS_FROM_WARNINGS="true")

devtools::install_github('cole-trapnell-lab/leidenbase')
devtools::install_github('cole-trapnell-lab/monocle3')

```

# Obtaining Dataset

For this walkthrough, you will need an example dataset from 10X Genomics. 
You can download that dataset 
(here)[https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz]
Decompress that file after you download it.

From the command line you can obtain the dataset and decompress it as follows:

```{bash eval = FALSE}
# if you have wget installed
wget https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz

# if you don't
curl -o pbmc3k_filtered_gene_bc_matrices.tar.gz https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz

tar -zxvf pbmc3k_filtered_gene_bc_matrices.tar.gz
```

# Loading data
Now that we have some data, we need to load it into a data object that can be 
used by Seurat or Monocle.

```{r, eval=FALSE}

# Seurat
pbmc.data <- Read10X(data.dir = "filtered_gene_bc_matrices/hg19/")
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)

#monocle
cds <- load_mm_data(mat_path = "filtered_gene_bc_matrices/hg19/matrix.mtx", 
                    feature_anno_path = "filtered_gene_bc_matrices/hg19/genes.tsv", 
                    cell_anno_path = "filtered_gene_bc_matrices/hg19/barcodes.tsv")
colnames(rowData(cds)) <- "gene_short_name"

# this will fail if there are duplicate gene names
rownames(cds) <- rowData(cds)$gene_short_name

# filter as above
rc <- apply(counts(cds), 1, function(x) { sum(x > 0) })
cc <- apply(counts(cds), 2, function(x) { sum(x > 0) })
cds <- cds[which(rc >= 3), which(cc >= 200)]
```

You can also convert between the two formats.

```{r, eval=FALSE}

seuratToMonocle3 <- function(x, assay = NULL) {
  # based on as.CellDataSet function in Seurat, updated for Monocle3
  if(is.null(assay))
    assay <- DefaultAssay(x)

  counts <- GetAssayData(object = x, assay = assay, slot = "counts")
  cell_metadata <- x[[]]
  feature_metadata <- x[[assay]][[]]
  if (!"gene_short_name" %in% colnames(x = feature_metadata)) {
    feature_metadata$gene_short_name <- rownames(feature_metadata)
  }
  new_cell_data_set(counts,
                         cell_metadata = cell_metadata,
                         gene_metadata = feature_metadata)
}

cds <- seuratToMonocle3(pbmc)

# going the other way is simpler

pbmc2 <- CreateSeuratObject(counts = counts(cds), project = "pbmc3k", min.cells = 3, min.features = 200)
```

# QC And Preprocessing

We first need to normalize the data by adjusting for differences in sequencing 
depth and log transforming (after multiplying by 10,000)

```{r eval=FALSE}

pbmc <- NormalizeData(pbmc)

# for Monocle, normalization and dimension reduction occur in a single step.
# So this step will take a bit.
cds <- preprocess_cds(cds, num_dim = 100)

```

In the above, Monocle scaled and performed PCA along with normalization. We need 
to perform those steps separately in Seurat.

```{r eval=FALSE}

pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)
pbmc <- ScaleData(pbmc)
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
saveRDS(pbmc, "pbmc.rds")
saveRDS(cds, "cds.rds")

```

There are a variety of QC maneuvers you can consider. "Cells" with way too many 
genes may be doublets or aggregates. "Cells" with way too few genes may be 
empty droplets with just contaminating background RNA. We can visualize some of
these stats in a variety of ways.

```{r}
pbmc <- readRDS("pbmc.rds")
cds <- readRDS("cds.rds")

pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

```

I like to plot some other diagnostic plots including the "knee plot" and 
distribution of genes per cell and reads per cell. We can create some helper 
functions. Here we set up some standard themes and color schemes that we 
can use accross plots. 

```{r}
my_colors <- list(lightblue = "#9FAFCD",
                 yellow = "#E3E0B2")

scale_color_two <- function(..., values = c(my_colors$yellow, my_colors$lightblue))
  scale_color_manual(..., values = values)

rel_text <- function(l, x = 0.5, y = 0.8, fontface = 'plain', fontsize = '9', hjust = 0) {
  grob <- grobTree(textGrob(l, 
                            x = x,  
                            y = y, 
                            hjust = hjust,
                            gp = gpar(fontsize=fontsize, fontface=fontface)))
  annotation_custom(grob)
}

my_theme <- theme_bw() +
  theme(axis.title = element_text(face = 2, size=9),
        strip.background = element_blank(),
        strip.text = element_text(face = 2, size = 10),
        legend.text = element_text(size=10),
        legend.title = element_text(face = 2, size=10),
        legend.position = "bottom", 
        legend.margin=margin(0,0,0,0),
        legend.box.margin=margin(-7,0,0,0),
        plot.margin = margin(3, 1, 0, 1, "lines")) 

my_theme_nogrid <- theme(panel.grid = element_blank())

knee_plot <- function(data, rs = NULL, threshold = 1000) {
  if(!length(rs)) 
    rs <- rowSums(data)
  rs <- sort(rs, decreasing = TRUE)
  gd <- data.frame(x = log10(1:length(rs)), 
                   y = log10(rs),
                   color = factor(rs > threshold))
  lab.y <- max(gd$y)
  lab.x <- max(gd$x)
  ggplot(gd, aes(x=x, y=y, color=color)) +
    geom_point(size = 0.1) + 
    annotate(geom = "text", hjust = 1, 
             label = paste("Cells: ", sum(rs > threshold)), 
             x = lab.x, 
             y = lab.y) + 
    labs(color = "Is Cell") +
    ylab(bquote(bold("log"[10]~"UMI Count"))) +
    xlab(bquote(bold("log"[10]~"Rank"))) +
    scale_color_two() + 
    guides(color = guide_legend(override.aes = list(size = 5))) +
    my_theme
}

genes_per_cell <- function(data, gc = NULL, rs = NULL, threshold = 0, bins = 50) {
  if(!length(gc)) 
    gc <- rowSums(data != 0)
  if(!length(rs)) 
    rs <- rowSums(data)
  gc <- gc[which(rs > threshold)]
  gd <- data.frame(gc=gc)
  label <- paste0("Mean: ", round(mean(gc)), "\nCells: ", length(gc))

  
  ggplot(gd, aes(x=gc)) + 
    geom_histogram(bins = bins, 
                   fill = my_colors$lightblue, 
                   color = "#333333",
                   size = 0.1,
                   show.legend = FALSE) +
    xlab("Unique Genes Per Cell") +
    ylab("Number of Cells") +
    rel_text(label) +
    my_theme + 
    my_theme_nogrid 
}

umis_per_cell <- function(data, rs = NULL, threshold = 0, bins = 50) {
  if(!length(rs)) 
    rs <- rowSums(data)
  rs <- rs[which(rs > threshold)]
  gd <- data.frame(rs=rs)
  label <- paste0("Mean: ", round(mean(rs)), "\nCells: ", length(rs))
  
  ggplot(gd, aes(x=rs)) + 
    geom_histogram(bins = bins, 
                   fill = my_colors$yellow, 
                   color = "#333333",
                   size = 0.1,
                   show.legend = FALSE) +
    xlab("Unique Transcripts Per Cell") +
    ylab("Number of Cells") +
    rel_text(label) +
    my_theme + 
    my_theme_nogrid 
}

a <- genes_per_cell(normalized_counts(cds))
b <- umis_per_cell(normalized_counts(cds))
c <- knee_plot(normalized_counts(cds))
cowplot::plot_grid(a,b,c, nrow=1, labels=c("A", "B", "C"))
```
We can then perform further dimensional reduction and clustering via UMAP.

```{r}

#Seurat
pbmc <- RunUMAP(pbmc, dims=1:50)

pbmc <- FindNeighbors(pbmc, dims = 1:50)
pbmc <- FindClusters(pbmc, resolution = 0.01)
DimPlot(pbmc, reduction = "umap")

#monocle
cds <- reduce_dimension(cds) # umap by default
cds = cluster_cells(cds, resolution=1e-5)
plot_cells(cds, color_cells_by = "cluster")
```

And perform differential gene expression analysis

```{r, warning=FALSE, message=FALSE, error=FALSE, results='hide', fig.keep='all'}

# Seurat
cluster2.markers <- FindMarkers(pbmc, ident.1 = 2, min.pct = 0.25)
head(cluster2.markers, n = 20)

# monocle3

marker_test_res <- top_markers(cds, group_cells_by="cluster", 
                               reference_cells=1000, cores=8)

top_specific_markers <- marker_test_res %>%
                            filter(fraction_expressing >= 0.10) %>%
                            group_by(cell_group) %>%
                            top_n(10, pseudo_R2)

top_specific_marker_ids <- unique(top_specific_markers %>% pull(gene_id))
plot_genes_by_group(cds,
                    top_specific_marker_ids,
                    group_cells_by="cluster",
                    ordering_type="maximal_on_diag",
                    max.size=3)
```

Are the genes identified by the two packages the same? Pretty close!

```{r}
ix <- which(top_specific_markers$cell_group == 3)
cl3_monocle <- top_specific_markers$gene_short_name[ix]

# how many significant genes as defined by Seurat are in top 10 
# as defined by monocle?

ix <- which(cluster2.markers$p_val_adj < 0.001)
cl2_seurat <- rownames(cluster2.markers)[ix]
sum(cl3_monocle %in% cl2_seurat)

# how many of the top 10 most significant genes as defined by Seurat are in 
# top 10 as defined by monocle?
ix <- head(order(cluster2.markers$p_val), n=10)
cl2_seurat <- rownames(cluster2.markers)[ix]
sum(cl3_monocle %in% cl2_seurat)

# 70% ! Not bad!
```

We can also identify genes associated with other metadata we have concerning
the cells (diagnosis, outcome, etc.). Here we will classify cells based on 
gene markers and use those classifications for differential gene expression. 
(Unfortunately, there seems to be a bug in monocle that is interfering with 
extraction of the resulting gene level coefficients)

```{r}

type <- rep("Unknown", ncol(cds))
exp <- normalized_counts(cds)
ix.b <- which(exp["CD3E", ] == 0 & exp["CD19",] > 0)
ix.cd4 <- which(exp["CD3E", ] > 0 & exp["CD4",] > 0)
ix.cd8 <- which(exp["CD3E", ] > 0 & exp["CD8A",] > 0)
ix.nk <- which(exp["CD3E", ] == 0 & exp["NCAM1",] > 0)
ix.mon <- which(exp["CD3E", ] == 0 & exp["CD14",] > 0)
type[ix.b] <- "B"
type[ix.cd4] <- "CD4+ T"
type[ix.cd8] <- "CD8+ T"
type[ix.nk] <- "Natural Killer"
type[ix.mon] <- "Monocytes"
colData(cds)$CellType <- type
ix.unk <- which(type == "Unknown")
plot_cells(cds[ , -ix.unk], color_cells_by = "CellType",  
           alpha = 0.5,
           label_cell_groups = FALSE,
           show_trajectory_graph = FALSE,
           cell_size = 1)

# gene_fits <- fit_models(cds[ , -ix.unk], model_formula_str = "~ CellType")
#fit_coefs <- coefficient_table(gene_fits)
#fit_coefs %>% filter(term != "(Intercept)") %>%
#      select(gene_short_name, term, q_value, estimate)

```

# Pseudotime

Pseudotime (ordering cells based on their gene expression distance from 
one another) is a core function of monocle.

```{r}
cds <- learn_graph(cds)
cds <- order_cells(cds, root_pr_nodes = "Y_1")
plot_cells(cds,
           color_cells_by = "pseudotime")

plot_genes_in_pseudotime(cds[cl3_monocle[1:5],], min_expr=0.5)

```

# Further reading

[Seurat demos](https://satijalab.org/seurat/vignettes.html)

[Monocle tutorial](https://cole-trapnell-lab.github.io/monocle3/)

[RNA Velocity Demo](https://combine-lab.github.io/alevin-tutorial/2020/alevin-velocity/)

[RNA Velocity Demo](https://vanandelinstitute.github.io/single_cell_analyses/rnavel_alevin_scvelo.html)

[A real world example](https://github.com/VanAndelInstitute/va_ecls)
